finagle-netty4/src/test/scala/com/twitter/finagle/netty4/Netty4SslTest.scala

```diff
@@ -0,0 +1,127 @@
+import com.twitter.finagle.ssl.{ClientAuth, Engine}
+import com.twitter.finagle.transport.Transport
+import com.twitter.finagle.transport.Transport.ServerSsl
+import com.twitter.util.{Await, Future}
+import io.netty.channel.ChannelPipeline
+import io.netty.handler.codec.string.{StringDecoder, StringEncoder}
+import io.netty.handler.codec.{DelimiterBasedFrameDecoder, Delimiters}
+import io.netty.handler.ssl.SslContextBuilder
+import io.netty.handler.ssl.util.SelfSignedCertificate
+import java.net.{InetAddress, InetSocketAddress, SocketAddress}
+import java.nio.charset.StandardCharsets.UTF_8
+import org.scalatest.Outcome
+import org.scalatest.fixture.FunSuite
+import org.scalatest.concurrent.{Eventually, IntegrationPatience}
+
+object Netty4SslTest {
+  case class Client(clientCert: SelfSignedCertificate, serverCert: SelfSignedCertificate,
+      params: Params = Params.empty,
+      stack: Stack[ServiceFactory[String, String]] = StackClient.newStack)
+    extends StdStackClient[String, String, Client] {
+
+    override protected type In = String
+    override protected type Out = String
+
+    override protected def newTransporter(addr: SocketAddress): Transporter[String, String] =
+      Netty4Transporter.raw[String, String](pipeline => {
+        pipeline.addLast(new DelimiterBasedFrameDecoder(8192, Delimiters.lineDelimiter(): _*))
+        pipeline.addLast(new StringDecoder())
+        pipeline.addLast(new StringEncoder())
+      }, addr, params)
+
+    override protected def newDispatcher(transport: Transport[String, String]): Service[String, String] =
+      new SerialClientDispatcher(transport)
+
+    override protected def copy1(stack: Stack[ServiceFactory[String, String]], params: Params) =
+      copy(clientCert, serverCert, params, stack)
+  }
+}
+
+class Netty4SslTest extends FunSuite with Eventually with IntegrationPatience {
+
+  class Ctx {
+    val serverCert = new SelfSignedCertificate("example.server.com")
+    val clientCert = new SelfSignedCertificate("example.client.com")
+    val allocator = io.netty.buffer.UnpooledByteBufAllocator.DEFAULT
+
+    private object StringServerInit extends (ChannelPipeline => Unit) {
+      def apply(pipeline: ChannelPipeline): Unit = {
+        pipeline.addLast("line", new DelimiterBasedFrameDecoder(100, Delimiters.lineDelimiter(): _*))
+        pipeline.addLast("stringDecoder", new StringDecoder(UTF_8))
+        pipeline.addLast("stringEncoder", new StringEncoder(UTF_8))
+      }
+    }
+
+    val server = {
+      val service =
+        new Service[String, String] {
+          override def apply(request: String): Future[String] = {
+            Future.value(
+              Transport.peerCertificate match {
+                case Some(_) => "OK\n"
+                case None => "ERROR\n"
+              }
+            )
+          }
+        }
+
+      val p = Params.empty +
+        ServerSsl(Some(SslServerConfiguration(clientAuth = ClientAuth.Needed))) +
```

Comment:
Please change to use the Netty4ServerEngineFactory instead. Something along the lines of

val config = SslServerConfiguration(
  keyCredentials = KeyCredentials.CertAndKey(new File(serverCert.cert()), new File(serverCert.key()),
  trustCredentials = TrustCredentials.CertCollection(new File(clientCert.cert()),
  clientAuth = ClientAuth.Needed)

...